src/libopensc/card-esteid2018.c

```diff
@@ -0,0 +1,347 @@
+static struct sc_card_operations esteid_ops;
+
+static struct sc_card_driver esteid2018_driver = {"EstEID 2018", "esteid2018", &esteid_ops, NULL, 0, NULL};
+
+struct esteid_priv_data {
+	sc_security_env_t sec_env; /* current security environment */
+};
+
+#define DRVDATA(card) ((struct esteid_priv_data *)((card)->drv_data))
+
+static int esteid_match_card(sc_card_t *card) {
+	int i = 0;
+	int r = 0;
+
+	i = _sc_match_atr(card, esteid_atrs, &card->type);
+	if (i >= 0) {
+		r = gp_select_aid(card, &IASECC_AID);
+		if (r == SC_SUCCESS) {
+			card->name = esteid_atrs[i].name;
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static int esteid_check_sw(sc_card_t *card, unsigned int sw1, unsigned int sw2) {
+	if (sw1 == 0x6B && sw2 == 0x00)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_FILE_END_REACHED);
+	return iso_ops->check_sw(card, sw1, sw2);
+}
+
+static int esteid_select(struct sc_card *card, unsigned char p1, unsigned char id1, unsigned char id2) {
+	struct sc_apdu apdu;
+	unsigned char sbuf[2];
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	// Select EF/DF
+	sbuf[0] = id1;
+	sbuf[1] = id2;
+
+	sc_format_apdu(card, &apdu, SC_APDU_CASE_1, 0xA4, p1, 0x0C);
+	if (id1 != 0x3F && id2 != 0x00) {
+		apdu.cse = SC_APDU_CASE_3_SHORT;
+		apdu.lc = 2;
+		apdu.data = sbuf;
+		apdu.datalen = 2;
+	}
+	apdu.le = 0x00;
+	apdu.resplen = 0;
+
+	LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");
+	LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), "SELECT failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+static int esteid_select_file(struct sc_card *card, const struct sc_path *in_path, struct sc_file **file_out) {
+	unsigned char pathbuf[SC_MAX_PATH_SIZE], *path = pathbuf;
+	int pathlen;
+	struct sc_file *file = NULL;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	// Only support full paths
+	if (in_path->type != SC_PATH_TYPE_PATH) {
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	}
+
+	memcpy(path, in_path->value, in_path->len);
+	pathlen = in_path->len;
+
+	while (pathlen >= 2) {
+		if (memcmp(path, "\x3F\x00", 2) == 0) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x00, 0x3F, 0x00), "MF select failed");
+		} else if (pathlen >= 2 && path[0] == 0xAD) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x01, path[0], path[1]), "DF select failed");
+		} else if (pathlen == 2) {
+			LOG_TEST_RET(card->ctx, esteid_select(card, 0x02, path[0], path[1]), "EF select failed");
+
+			if (file_out != NULL) // Just make a dummy file
+			{
+				file = sc_file_new();
+				if (file == NULL)
+					LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+				file->path = *in_path;
+
+				*file_out = file;
+			}
+		}
+		path += 2;
+		pathlen -= 2;
+	}
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+// temporary hack, overload 6B00 SW processing
+static int esteid_read_binary(struct sc_card *card, unsigned int idx, u8 *buf, size_t count, unsigned long flags) {
+	int r;
+	LOG_FUNC_CALLED(card->ctx);
+	void *saved = card->ops->check_sw;
```

Comment:
typedef int (*check_sw)(struct sc_card *card,unsigned int sw1,unsigned int sw2);
	check_sw saved = card->ops->check_sw;

Reply:
Correct me if I'm wrong, but handling a void pointer in this context should be sufficient effectively (a pointer is a pointer no matter where it points)

Reply:
https://stackoverflow.com/questions/36645660/why-cant-i-cast-a-function-pointer-to-void